src/Plutarch/Test/Export.hs

```diff
@@ -0,0 +1,728 @@
+  where
+    go :: WithExportEnv dat red info -> WithExportEnv dat red info
+    go env =
+      let collected =
+            ScriptParams
+              (view #datum env)
+              (view #redeemer env)
+              (view #context env)
+          modified = f (view #info env) collected
+       in set #datum (view #datum modified)
+            . set #redeemer (view #redeemer modified)
+            . set #context (view #context modified)
+            $ env
+
+{- | As 'modifyingScriptParams', except the \'locally scoped\' arguments are replaced with
+ the given ones.
+
+ @since 1.2.1
+-}
+settingScriptParams ::
+  forall (dat :: Type) (red :: Type) (info :: Type) (a :: Type).
+  -- | How to \'locally set\' the script arguments
+  (info -> ScriptParams dat red) ->
+  -- | The collection of tests to run using the set arguments
+  WithExport dat red info a ->
+  WithExport dat red info a
+settingScriptParams f = modifyingScriptParams (\i _ -> f i)
+
+{- | As 'modifyingScriptParams', but without additional information from the script
+ export.
+
+ @since 1.2.1
+-}
+modifyingScriptParams' ::
+  forall (dat :: Type) (red :: Type) (info :: Type) (a :: Type).
+  -- | How to \'locally modify\' the arguments
+  (ScriptParams dat red -> ScriptParams dat red) ->
+  -- | The collection of tests to run using the modified arguments
+  WithExport dat red info a ->
+  WithExport dat red info a
+modifyingScriptParams' f = modifyingScriptParams (\_ x -> f x)
+
+{- | As 'settingScriptParams', but without additional information from the script
+ export.
+
+ @since 1.2.1
+-}
+settingScriptParams' ::
+  forall (dat :: Type) (red :: Type) (info :: Type) (a :: Type).
+  -- | A \'local value\' for the arguments
+  ScriptParams dat red ->
+  -- | The collection of tests to run using the set arguments
+  WithExport dat red info a ->
+  WithExport dat red info a
+settingScriptParams' x = modifyingScriptParams (\_ _ -> x)
+
+-- Helpers
+
+runWithExport ::
+  forall (dat :: Type) (red :: Type) (info :: Type) (param :: Type).
+  WithExport dat red info () ->
+  RawScriptExport ->
+  Linker param (ScriptExport info) ->
+  TestParams dat red param ->
+  Either LinkerError (Map Text PreparedScriptTests)
+runWithExport (WithExport comp) raw linker args =
+  case runLinker linker raw . view #linkerParams $ args of
+    Left err -> Left err
+    Right linked ->
+      let env =
+            WithExportEnv
+              (view (#scriptParams %% #datum) args)
+              (view (#scriptParams %% #redeemer) args)
+              (view (#scriptParams %% #context) args)
+              raw
+              (view #information linked)
+       in Right . coerce . snd . execRWS comp env $ ()
+
+-- Note from Koz: the manual lenses are due to a weird interaction between TH
+-- and record syntax. I wasn't able to establish _why_ it was being weird, but
+-- when written this way, it behaves.
+data WithExportEnv (dat :: Type) (red :: Type) (info :: Type)
```

Comment:
Would it be a good idea to open issue on optics repo? I'll make one if so

Reply:
I'm unsure if this is an `optics` issue per se. It could be something unrelated: that's just my best guess. If I have time and nothing else to do I can try and reproduce it minimally, but for now, we can just go with what works.